### ISSUE TITLE
SSLError: [Errno 24] Too many open files

### ISSUE THREAD

**Issue by francoislaurent** [francoislaurent/escale#19]
In version 0.4rc2:
```
2/05 17:32 syncacre.*[CRITICAL]	manager.run:135	Traceback (most recent call last):
  File "/home/larva/syncacre/syncacre/manager/manager.py", line 112, in run
    new |= self.download()
  File "/home/larva/syncacre/syncacre/manager/manager.py", line 189, in download
    ok = self.relay.pop(remote_file, temp_file, blocking=False, **self.pop_args)
  File "/home/larva/syncacre/syncacre/relay/relay.py", line 568, in pop
    self.markAsRead(remote_file, **kwargs)
  File "/home/larva/syncacre/syncacre/relay/relay.py", line 591, in markAsRead
    self._push(local_placeholder, remote_placeholder)
  File "/home/larva/syncacre/syncacre/relay/webdav.py", line 306, in _push
    self.webdav.upload(local_file, remote_file)
  File "/home/larva/syncacre/syncacre/relay/webdav.py", line 143, in upload
    _easywebdav_adapter(self._upload, 'rb', local_path_or_fileobj, remote_path)
  File "/home/larva/syncacre/syncacre/relay/webdav.py", line 50, in _easywebdav_adapter
    fun(f, *args)
  File "/home/larva/.local/lib/python2.7/site-packages/easywebdav/client.py", line 160, in _upload
    self._send('PUT', remote_path, (200, 201, 204), data=fileobj)
  File "/home/larva/syncacre/syncacre/relay/webdav.py", line 90, in _send
    return easywebdav.Client._send(self, *args, **kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/easywebdav/client.py", line 97, in _send
    response = self.session.request(method, url, allow_redirects=False, **kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/requests/sessions.py", line 488, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/requests/sessions.py", line 609, in send
    r = adapter.send(request, **kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/requests/adapters.py", line 497, in send
    raise SSLError(e, request=request)
SSLError: [Errno 24] Too many open files
```
To be confirmed in later versions.

At the same time, the webdav server was displaying "Your storage is full, files can not be updated or synced anymore!"

**Comment by francoislaurent** [comment]
Confirmed in 0.4:
```
03/05 21:47 syncacre.*[CRITICAL]  manager.run:135 Traceback (most recent call last):
  File "/home/larva/syncacre/syncacre/manager/manager.py", line 112, in run
    new |= self.download()
  File "/home/larva/syncacre/syncacre/manager/manager.py", line 189, in download
    ok = self.relay.pop(remote_file, temp_file, blocking=False, **self.pop_args)
  File "/home/larva/syncacre/syncacre/relay/relay.py", line 549, in pop
    if not self.acquireLock(remote_file, blocking=blocking):
  File "/home/larva/syncacre/syncacre/relay/relay.py", line 465, in acquireLock
    self.touch(self.lock(remote_file), content=self.client)
  File "/home/larva/syncacre/syncacre/relay/relay.py", line 399, in touch
    self._push(local_file, remote_file)
  File "/home/larva/syncacre/syncacre/relay/webdav.py", line 306, in _push
    self.webdav.upload(local_file, remote_file)
  File "/home/larva/syncacre/syncacre/relay/webdav.py", line 143, in upload
    _easywebdav_adapter(self._upload, 'rb', local_path_or_fileobj, remote_path)
  File "/home/larva/syncacre/syncacre/relay/webdav.py", line 50, in _easywebdav_adapter
    fun(f, *args)
  File "/home/larva/.local/lib/python2.7/site-packages/easywebdav/client.py", line 160, in _upload
    self._send('PUT', remote_path, (200, 201, 204), data=fileobj)
  File "/home/larva/syncacre/syncacre/relay/webdav.py", line 90, in _send
    return easywebdav.Client._send(self, *args, **kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/easywebdav/client.py", line 97, in _send
    response = self.session.request(method, url, allow_redirects=False, **kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/requests/sessions.py", line 488, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/requests/sessions.py", line 609, in send
    r = adapter.send(request, **kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/requests/adapters.py", line 497, in send
    raise SSLError(e, request=request)
SSLError: [Errno 24] Too many open files
```

**Comment by francoislaurent** [comment]
Confirmed in 0.4.1rc1:
```
04/05 06:41 syncacre.*[CRITICAL]  manager.run:148 Traceback (most recent call last):
  File "/home/larva/syncacre/syncacre/manager/manager.py", line 123, in run
    new |= self.download()
  File "/home/larva/syncacre/syncacre/manager/manager.py", line 190, in download
    placeholder = self.relay.getPlaceholder(remote_file)
  File "/home/larva/syncacre/syncacre/relay/relay.py", line 435, in getPlaceholder
    self._get(remote_placeholder, local_placeholder)
  File "/home/larva/syncacre/syncacre/relay/webdav.py", line 332, in _get
    self.webdav.download(remote_file, local_file)
  File "/home/larva/syncacre/syncacre/relay/webdav.py", line 158, in download
    response = self._send('GET', remote_path, 200, stream=True)
  File "/home/larva/syncacre/syncacre/relay/webdav.py", line 91, in _send
    return easywebdav.Client._send(self, *args, **kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/easywebdav/client.py", line 97, in _send
    response = self.session.request(method, url, allow_redirects=False, **kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/requests/sessions.py", line 488, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/requests/sessions.py", line 609, in send
    r = adapter.send(request, **kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/requests/adapters.py", line 497, in send
    raise SSLError(e, request=request)
SSLError: [Errno 24] Too many open files
```

**Comment by francoislaurent** [comment]
SSLError is supposed to be raised by OpenSSL. Local OpenSSL version is 1.0.2g.

requests.Session is supposed to reuse TCP connections. However it is not clear whether it reuses an ssl.SSLContext.

Could this be similar to an issue with the FTP backend and vsftpd that I solved adding 'require_ssl_reuse=NO' to the '/etc/vsftpd.conf' file?

**Comment by francoislaurent** [comment]
An idea from https://docs.python.org/2/library/socket.html#example:
> ```
> socket.error: [Errno 98] Address already in use
> ```
> 
> This is because the previous execution has left the socket in a TIME_WAIT state, and can’t be immediately reused.
> 
> There is a socket flag to set, in order to prevent this, socket.SO_REUSEADDR:
> 
> ```
> s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
> s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
> s.bind((HOST, PORT))
> ```
> 
> the SO_REUSEADDR flag tells the kernel to reuse a local socket in TIME_WAIT state, without waiting for its natural timeout to expire.

**Comment by francoislaurent** [comment]
Still here in 0.4.1:
```
05/05 08:37 syncacre.*[CRITICAL]  manager.run:151 Traceback (most recent call last):
  File "/home/larva/syncacre/syncacre/manager/manager.py", line 126, in run
    new |= self.download()
  File "/home/larva/syncacre/syncacre/manager/manager.py", line 193, in download
    placeholder = self.relay.getPlaceholder(remote_file)
  File "/home/larva/syncacre/syncacre/relay/relay.py", line 435, in getPlaceholder
    self._get(remote_placeholder, local_placeholder)
  File "/home/larva/syncacre/syncacre/relay/webdav.py", line 371, in _get
    self.webdav.download(remote_file, local_file)
  File "/home/larva/syncacre/syncacre/relay/webdav.py", line 183, in download
    response = self._send('GET', remote_path, 200, stream=True)
  File "/home/larva/syncacre/syncacre/relay/webdav.py", line 106, in _send
    self._response = easywebdav.Client._send(self, *args, **kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/easywebdav/client.py", line 97, in _send
    response = self.session.request(method, url, allow_redirects=False, **kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/requests/sessions.py", line 488, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/requests/sessions.py", line 609, in send
    r = adapter.send(request, **kwargs)
  File "/home/larva/.local/lib/python2.7/site-packages/requests/adapters.py", line 497, in send
    raise SSLError(e, request=request)
SSLError: [Errno 24] Too many open files
```
Last fix fails to catch SSLError exceptions.

**Comment by francoislaurent** [comment]
``-r`` option offers a working workaround for single client instances.

The error is still not confirmed on Python 3 and with multi-client instances.

**Comment by francoislaurent** [comment]
Confirmed in Python 3.

**Comment by francoislaurent** [comment]
In multiclient mode:
```
25/05 20:40 [CRITICAL]     manager.run:192 Traceback (most recent call last):
  File "*/manager/manager.py", line 169, in run
    new |= self.download()
  File "*/manager/manager.py", line 264, in download
    ok = self.relay.pop(remote_file, temp_file, blocking=False, **self.pop_args)
  File "*/relay/relay.py", line 730, in pop
    if not self.acquireLock(remote_file, mode='r', blocking=blocking):
  File "*/relay/relay.py", line 643, in acquireLock
    self.touch(self.lock(remote_file), content=repr(lock_info))
  File "*/relay/relay.py", line 505, in touch
    with open(local_file, 'w') as f:
OSError: [Errno 24] Too many open files: '/tmp/tmplgjj63v8'

25/05 20:53 [CRITICAL]     controller.failure:73   UnrecoverableError: [Errno 24] Too many open files
25/05 20:53 [CRITICAL]     controller.failure:76   the Python environment should be reset
```

**Comment by francoislaurent** [comment]
```
27/05 07:26 [CRITICAL]     manager.run:192 Traceback (most recent call last):
  File "*/manager/manager.py", line 169, in run
  File "*/manager/manager.py", line 264, in download
  File "*/relay/relay.py", line 731, in pop
  File "*/relay/relay.py", line 644, in acquireLock
  File "*/relay/relay.py", line 506, in touch
IOError: [Errno 24] Too many open files: '/tmp/tmpWBnKUf'

27/05 07:26 [CRITICAL]     controller.failure:73   UnrecoverableError: [Errno 24] Too many open files
27/05 07:26 [CRITICAL]     controller.failure:76   the Python environment should be reset
```

**Comment by francoislaurent** [comment]
```
28/05 01:51 [CRITICAL]     manager.run:176 Traceback (most recent call last):
  File "*/.local/lib/python3.5/site-packages/requests/packages/urllib3/util/ssl_.py", line 309, in ssl_wrap_socket
    context.load_verify_locations(ca_certs, ca_cert_dir)
OSError: [Errno 24] Too many open files

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "*/.local/lib/python3.5/site-packages/requests/adapters.py", line 438, in send
    timeout=timeout
  File "*/.local/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "*/.local/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 345, in _make_request
    self._validate_conn(conn)
  File "*/.local/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 844, in _validate_conn
    conn.connect()
  File "*/.local/lib/python3.5/site-packages/requests/packages/urllib3/connection.py", line 326, in connect
    ssl_context=context)
  File "*/.local/lib/python3.5/site-packages/requests/packages/urllib3/util/ssl_.py", line 311, in ssl_wrap_socket
    raise SSLError(e)
requests.packages.urllib3.exceptions.SSLError: [Errno 24] Too many open files

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "*/manager/manager.py", line 153, in run
    new |= self.download()
  File "*/manager/manager.py", line 254, in download
    ok = self.relay.pop(remote_file, temp_file, blocking=False, **self.pop_args)
  File "*/relay/relay.py", line 752, in pop
    self.markAsRead(remote_file, **kwargs)
  File "*/relay/relay.py", line 781, in markAsRead
    self._push(local_placeholder, remote_placeholder)
  File "*/relay/webdav.py", line 407, in _push
    self.webdav.upload(local_file, remote_file)
  File "*/relay/webdav.py", line 205, in upload
    self._upload(f, remote_path)
  File "*/.local/lib/python3.5/site-packages/easywebdav/client.py", line 160, in _upload
    self._send('PUT', remote_path, (200, 201, 204), data=fileobj)
  File "*/relay/webdav.py", line 162, in _send
    raise e
  File "*/relay/webdav.py", line 118, in _send
    self._response = self.__send(*args, **kwargs)
  File "*/relay/webdav.py", line 103, in __send
    response = self.session.request(method, url, allow_redirects=False, **kwargs)
  File "*/.local/lib/python3.5/site-packages/requests/sessions.py", line 518, in request
    resp = self.send(prep, **send_kwargs)
  File "*/.local/lib/python3.5/site-packages/requests/sessions.py", line 639, in send
    r = adapter.send(request, **kwargs)
  File "*/.local/lib/python3.5/site-packages/requests/adapters.py", line 512, in send
    raise SSLError(e, request=request)
requests.exceptions.SSLError: [Errno 24] Too many open files
```

**Comment by francoislaurent** [comment]
Confirmed as solved in 0.5rc1.

Beyond some fixes in the WebDAV backend that consist of closing streamed responses, the bug was largely due to deficiencies in the relay and encryption logic.

mkstemp was called to get a temporary file name and no care was taken to close an open file descriptor returned by mkstemp. These descriptors were not garbage collected.
